src/mds/events/ESession.h

```diff
@@ -40,9 +42,11 @@ class ESession : public LogEvent {
     client_inst(inst), open(o), cmapv(v), inotablev(0),
     client_metadata(cm) { }
   ESession(const entity_inst_t& inst, bool o, version_t v,
-	   const interval_set<inodeno_t>& i, version_t iv) :
+	   const interval_set<inodeno_t>& i, version_t iv,
```

Comment:
change to `interval_set<inodeno_t> i`

and then

`inos(std::move(i))`

It's better practice to pass the object by value if you're going to just do a copy anyway. This let's the caller choose to move the object to avoid the copy.

Reply:
I think using semantics of move obj has side effects in constructors. The parameters will lose object ownership.

like: 
Server::journal_close_session{
...
interval_set<inodeno_t> both;
LogEvent* e = new ESession(session->info.inst, false, pv, both, piv, session->info.prealloc_inos);                                                   
 MDSLogContextBase* c = new C_MDS_session_finish(this, session, sseq, false, pv, both, piv,                                                           
                                                  session->info.prealloc_inos, ls, on_safe);
}
the both will lost data after create ESession. I think.

Reply:
No, the interval_set will be copied when passed by value to the `ESession` constructor.

Reply:
I checkout the constructor of ESession.
the interval_set is passed  by reference:
like this:
   ESession(const entity_inst_t& inst, bool o, version_t v,
-          const interval_set<inodeno_t>& i, version_t iv) :
+          interval_set<inodeno_t>& i, version_t iv,
+          interval_set<inodeno_t>& _purge_inos) :
     LogEvent(EVENT_SESSION),
-    client_inst(inst), open(o), cmapv(v), inos(i), inotablev(iv) { }
+    client_inst(inst), open(o), cmapv(v), inos(i), inotablev(iv),
+    purge_inos(_purge_inos) {}

Reply:
@batrick  I think pass-by-value may not be not as or more efficient than pass-by-reference for lvalues, we always need to copy it once.
If we need to consider possible scenarios, pass by rvalues ,in the future, I will update it.

Reply:
No, pass by value is preferred. So literally:

`ESession(const entity_inst_t& inst, bool, o, version_t v, interval_set<inodeno_t> i, version_t iv, interval_set<inodeno_t> _purge_inos)`

See also this link for rationale: https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference

Reply:
updated it.